### PR TITLE
add utils to ddDrakeModel

### PR DESCRIPTION
**Pull request by manuelli**
- Add some utilities to ddDrakeModel.
- Add utilities to python side of director.

@patmarion could you review?

### CODE REVIEW

File: src/app/ddDrakeModel.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <QObject>
 #include <QColor>
 #include <QVector>
+#include <QGenericMatrix>
```
Comment (patmarion) [inline src/app/ddDrakeModel.h]: are you using this?

File: src/app/wrapped_methods_drake.txt [hunk 2]
```diff
@@ -12,11 +12,17 @@ QVector<double> ddDrakeModel::getJointPositions() const;
...
 bool ddDrakeModel::getLinkToWorld(const QString&, vtkTransform*);
+bool ddDrakeModel::getFrameToWorld(int, vtkTransform*);
 QString ddDrakeModel::getLinkNameForMesh(vtkPolyData*);
 QStringList ddDrakeModel::getLinkNames();
 QStringList ddDrakeModel::getJointNames();
+QString ddDrakeModel::getBodyOrFrameName(int);
 int ddDrakeModel::findLinkID(const QString&) const;
+int ddDrakeModel::findJointID(const QString&) const;
+int ddDrakeModel::findFrameID(const QString&) const;
+QString ddDrakeModel::FindNameOfChildBodyOfJoint(const QString &) const;
```
Comment (patmarion) [inline src/app/wrapped_methods_drake.txt]: lowercase

File: src/python/director/ioUtils.py [hunk 3]
```diff
@@ -147,3 +149,17 @@ def _triangulate(polyData):
     normals.SetInput(polyData)
     normals.Update()
     return shallowCopy(normals.GetOutput())
+
+
+def saveDataToFile(filename, dataDict, overwrite=False):
+    if overwrite is False and os.path.isfile(filename):
+        raise ValueError("file already exists, overwrite option was False")
+
+    my_shelf = shelve.open(filename,'n')
```
Comment (patmarion) [inline src/python/director/ioUtils.py]: myShelf

File: src/python/director/roboturdf.py [hunk 4]
```diff
@@ -231,10 +239,11 @@ def loadRobotModel(name=None, view=None, parent='scene', urdfFile=None, color=No
 
     return obj, jointController
 
-
-def loadRobotModelFromFile(filename):
+# floating joint type can be one of 'FIXED','ROLLPITCHYAW','QUATERNION'
+def loadRobotModelFromFile(filename, floating_joint_type='ROLLPITCHYAW'):
```
Comment (patmarion) [inline src/python/director/roboturdf.py]: camelCase

File: src/python/director/roboturdf.py [hunk 5]
```diff
@@ -231,10 +239,11 @@ def loadRobotModel(name=None, view=None, parent='scene', urdfFile=None, color=No
...
     model = PythonQt.dd.ddDrakeModel()
-    if not model.loadFromFile(filename):
+    print ("Python: floating joint type = " + floating_joint_type)
```
Comment (patmarion) [inline src/python/director/roboturdf.py]: no print

File: src/python/director/transformUtils.py [hunk 6]
```diff
@@ -167,6 +168,8 @@ def transformFromPose(position, quaternion):
     '''
     Returns a vtkTransform
     '''
+    rotationMatrix = np.zeros((3,3))
+    vtk.vtkMath.QuaternionToMatrix3x3(quaternion, rotationMatrix)
```
Comment (patmarion) [inline src/python/director/transformUtils.py]: i think you can remove these two lines

File: src/python/director/transformUtils.py [hunk 7]
```diff
@@ -205,3 +208,41 @@ def copyFrame(transform):
...
+def forceMomentTransformation(inputFrame, outputFrame):
+    FM = np.zeros((6,6))
+
+    inputToOutputFrame = copyFrame(inputFrame)
+    inputToOutputFrame.PostMultiply()
+    inputToOutputFrame.Concatenate(outputFrame.GetLinearInverse())
+
+    position, quaternion = poseFromTransform(inputToOutputFrame)
+    inputToOutputRotationMatrix = np.zeros((3,3))
+    vtk.vtkMath.QuaternionToMatrix3x3(quaternion, inputToOutputRotationMatrix)
```
Comment (patmarion) [inline src/python/director/transformUtils.py]: use transformations library.  something like inputToOutputRotationMatrix = transformations.quaternion_matrix(quaternion)[:3,:3]

File: src/python/director/transformUtils.py [hunk 8]
```diff
@@ -205,3 +208,41 @@ def copyFrame(transform):
...
+    FM[3:,3:] = inputToOutputRotationMatrix
+
+    cross = crossProductMatrix(position)
+    
+    FM[0:3,3:] = np.dot(cross,inputToOutputRotationMatrix)
+
+    return FM
+
+# computes matrix P such that P*y = cross product of x and y
+def crossProductMatrix(x):
```
Comment (patmarion) [inline src/python/director/transformUtils.py]: move comment into doc string

Comment (patmarion) [inline src/python/director/transformUtils.py]: probably an existing numpy function to do this math

File: src/python/director/visualization.py [hunk 9]
```diff
@@ -6,12 +6,25 @@
 from director import callbacks
 from director import frameupdater
 from PythonQt import QtCore, QtGui
+import numpy as np
+
+
```
Comment (patmarion) [inline src/python/director/visualization.py]: remove from diff

File: src/python/director/visualization.py [hunk 10]
```diff
@@ -6,12 +6,25 @@
...
+def computeAToB(a,b):
+
+    t = vtk.vtkTransform()
+    t.PostMultiply()
+    t.Concatenate(b)
+    t.Concatenate(a.GetLinearInverse())
+    tt = vtk.vtkTransform()
+    tt.SetMatrix(t.GetMatrix())
+    return tt
+
```
Comment (patmarion) [inline src/python/director/visualization.py]: remove from diff

File: src/python/director/visualization.py [hunk 11]
```diff
@@ -928,6 +941,7 @@ def getRandomColor():
     return colorsys.hsv_to_rgb(np.random.rand(), 1.0, 0.9)
 
 
+
```
Comment (patmarion) [inline src/python/director/visualization.py]: remove

File: src/app/ddDrakeModel.cpp [hunk 12]
```diff
@@ -44,6 +46,13 @@
 #include <QMap>
 #include <QDir>
 
+#include <iostream>
+#include <fstream>
+#include <string>
```
Comment (patmarion) [inline src/app/ddDrakeModel.cpp]: check if you actually need these header

File: src/app/ddDrakeModel.cpp [hunk 13]
```diff
@@ -44,6 +46,13 @@
...
+
+
+
+
```
Comment (patmarion) [inline src/app/ddDrakeModel.cpp]: whitespace

File: src/app/ddDrakeModel.cpp [hunk 14]
```diff
@@ -1146,6 +1268,83 @@ void ddDrakeModel::getModelMesh(vtkPolyData* polyData)
...
 
+void ddDrakeModel::getModelMeshWithLinkInfoAndNormals(vtkPolyData* polyData)
+{
+  if (!polyData){
+    return;
+  }
+
+  vtkSmartPointer<vtkAppendPolyData> appendFilter = vtkSmartPointer<vtkAppendPolyData>::New();
+
+  for (const auto & rb: this->Internal->Model->bodies){
```
Comment (patmarion) [inline src/app/ddDrakeModel.cpp]: brace style

File: src/app/ddDrakeModel.cpp (hunk 12 above)
Comment (manuelli) [inline src/app/ddDrakeModel.cpp]: I removed these

File: src/app/ddDrakeModel.h (hunk 1 above)
Comment (manuelli) [inline src/app/ddDrakeModel.h]: removed

File: src/python/director/transformUtils.py (hunk 8 above)
Comment (manuelli) [inline src/python/director/transformUtils.py]: I looked around the numpy documentation but couldn't find a method that provides this functionality.

**Comment by manuelli** [comment]
I made the requested style changes. Am now doing some tests to make sure I haven't broken any functionality. Will let you know when I think it's ready for your final review.

**Comment by manuelli** [comment]
@patmarion  I made the requested changes, and touched up some other style discrepancies that I found. I wasn't able to find a method in numpy to make that cross product matrix, so I left that in. Let me know if you want me to make any more changes.

**Pull request merged**